### PR TITLE
implemented DateLessLogFormatHelper

### LINKED ISSUE #4697: Create a dateless helper for logs that lack a date for each timestamps
**Describe the problem:**

Some log files do not have the date in their timestamp. This is the case for ps.txt log files in MacOS/iOS for example. I wish to create a dateless_helper modeled on the yearless helper.

### PR DESCRIPTION
**Pull request by rick-slin**
## One line description of pull request

Implemented DateLessLogFormatHelper to retreive a date based on the modified, changed and creation time of a file.

## Description:
Implemented DateLessLogFormatHelper to retreive a date based on the modified, changed and creation time of a file.

**Related issue (if applicable):** fixes #4697 

## Notes:
All contributions to Plaso undergo [code review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc).
This makes sure that the code has appropriate test coverage and conforms to the
[Plaso style guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
* [ ] Automated checks (GitHub Actions, AppVeyor) pass
* [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
* [x] Reviewer assigned

### CODE REVIEW

File: plaso/containers/events.py [hunk 1]
```diff
@@ -82,6 +82,57 @@ def CalculateEventValuesHash(event_data, event_data_stream):
   return md5_context.hexdigest()
 
 
+class DateLessLogHelper(interface.AttributeContainer):
+  """Date-less log helper attribute container.
+
+  Attributes:
+    earliest_date (datetime.datetime): earliest possible date the event data
```
Comment (joachimmetz) [inline plaso/containers/events.py:89]: why use datetime, why not use a tuple of (year, month, day) ?

File: plaso/containers/events.py [hunk 2]
```diff
@@ -82,6 +82,57 @@ def CalculateEventValuesHash(event_data, event_data_stream):
...
+        stream was created.
+    latest_date (datetime.datetime): last relative date determined by the
+        date-less log helper.
+  """
+
+  CONTAINER_TYPE = 'date_less_log_helper'
+
+  SCHEMA = {
+    '_event_data_stream_identifier': 'AttributeContainerIdentifier',
```
Comment (joachimmetz) [inline plaso/containers/events.py:98]: style nit: use 4 space continuation indentation.

File: plaso/containers/events.py (hunk 1 above)
Comment (rick-slin) [reply plaso/containers/events.py:89]: I didn't think about it. It is a better idea.

**Comment by joachimmetz** [comment]
@rick-slin FYI I'll start looking at merging this after the March release I might combine it with the yearless log helper into 1

File: plaso/engine/timeliner.py [hunk 3]
```diff
@@ -75,6 +78,74 @@ def _CreateTimeZonePerPathSpec(self, system_configurations):
...
+
+    filter_expression = '_event_data_stream_identifier == "{0:s}"'.format(
+        lookup_key)
+    date_less_log_helpers = list(storage_writer.GetAttributeContainers(
+        events.DateLessLogHelper.CONTAINER_TYPE,
+        filter_expression=filter_expression))
+    if not date_less_log_helpers:
+      message = (
+          'missing date-less log helper, defaulting to current date: '
+          '{0:d}').format(self._current_date)
```
Comment (joachimmetz) [inline plaso/engine/timeliner.py:106]: Isn't current_data a datetime object?

File: plaso/engine/timeliner.py [hunk 4]
```diff
@@ -75,6 +78,74 @@ def _CreateTimeZonePerPathSpec(self, system_configurations):
...
+            'missing earliest and latest date in date-less log helper, '
+            'defaulting to current date: {0:d}').format(self._current_date)
+        self._ProduceTimeliningWarning(storage_writer, event_data, message)
+
+        base_date = self._current_date
+
+      elif earliest_date < self._current_date:
+        base_date = earliest_date
+
+      elif latest_date < self._current_year:
```
Comment (joachimmetz) [inline plaso/engine/timeliner.py:126]: Is this check as intended? given datetime is compared to int ?

Comment (joachimmetz) [reply plaso/engine/timeliner.py:126]: I'll assume this should be current_date

File: plaso/engine/timeliner.py [hunk 5]
```diff
@@ -183,16 +264,33 @@ def _GetEvent(
     """
     timestamp = None
     if date_time.is_delta:
-      base_year = self._GetBaseYear(storage_writer, event_data)
+      if (
+          date_time.year == 0 and
```
Comment (joachimmetz) [inline plaso/engine/timeliner.py:268]: what if the date-less log format spans multiple days?

File: plaso/engine/timeliner.py (hunk 4 above)
Comment (joachimmetz) [reply plaso/engine/timeliner.py:126]: @rick-slin can you confirm this should be the case

File: plaso/engine/timeliner.py [hunk 6]
```diff
@@ -183,16 +264,33 @@ def _GetEvent(
...
+          date_time.month == 0 and
```
Comment (joachimmetz) [inline plaso/engine/timeliner.py:269]: What if date_time returns None for the year, month and day_of_month ?

File: tests/engine/timeliner.py [hunk 7]
```diff
@@ -210,6 +262,25 @@ def testGetEvent(self):
     self.assertEqual(event.date_time.year, 4)
     self.assertEqual(event.timestamp, -62035818808570124)
 
+    # Test with the dateless helper
+    storage_writer = self._CreateStorageWriter(
+        event_data, base_date=datetime.datetime(2013, 7, 22))
+
+    date_time = dfdatetime_time_elements.TimeElementsInMicroseconds(
+        is_delta=True,
+        time_elements_tuple=(0, 0, 0, 20, 6, 31, 429876))
```
Comment (joachimmetz) [inline tests/engine/timeliner.py:271]: this is an invalid date time, month and day of month must start with 1

**Comment by joachimmetz** [comment]
@rick-slin unfortunately there are some shortcomings in the current design/approach, I'll need to take a closer look at a later time how to best implement support for date and time formats without a date.

**Comment by joachimmetz** [comment]
Closing in favor of https://github.com/log2timeline/plaso/pull/4851

File: tests/engine/timeliner.py (hunk 7 above)
Comment (joachimmetz) [reply tests/engine/timeliner.py:271]: dfDateTime changed to support delta times with month/day of month of 0

File: plaso/engine/timeliner.py (hunk 6 above)
Comment (joachimmetz) [reply plaso/engine/timeliner.py:269]: solved this differently by merging year-less and date-less helpers

File: plaso/engine/timeliner.py (hunk 5 above)
Comment (joachimmetz) [reply plaso/engine/timeliner.py:268]: No longer an issue in https://github.com/log2timeline/plaso/pull/4854

File: plaso/engine/timeliner.py (hunk 4 above)
Comment (joachimmetz) [reply plaso/engine/timeliner.py:126]: no longer an issue in https://github.com/log2timeline/plaso/pull/4854

File: plaso/engine/timeliner.py (hunk 3 above)
Comment (joachimmetz) [reply plaso/engine/timeliner.py:106]: No longer an issue in https://github.com/log2timeline/plaso/pull/4854

File: plaso/containers/events.py (hunk 1 above)
Comment (joachimmetz) [reply plaso/containers/events.py:89]: went for tuples in https://github.com/log2timeline/plaso/pull/4854

**Pull request closed without merging**